### PR TITLE
BUG: Fix test_from_object_array_unicode (test_defchararray.TestBasic)…

### PR DESCRIPTION
**Pull request by tadeu**
… on Windows debug

"isspace" crashes with an assert on debug when the character is not ASCII, in Windows

### CODE REVIEW

**Comment by charris** [comment]
This isn't actually a `TST`, that prefix is for official tests in `tests` directories.

File: numpy/core/src/multiarray/arrayobject.c [hunk 1]
```diff
@@ -880,7 +880,10 @@ _mystrncmp(char *s1, char *s2, int len1, int len2)
 
 #define SMALL_STRING 2048
 
-#if defined(isspace)
+/* defined(_WIN32) below is because MSVC "isspace" crashes with an assert on
+   debug when the character is not ASCII
+*/
+#if defined(isspace) || defined(_WIN32)
 #undef isspace
 #define isspace(c)  ((c==' ')||(c=='\t')||(c=='\n')||(c=='\r')||(c=='\v')||(c=='\f'))
```
Comment (charris) [inline numpy/core/src/multiarray/arrayobject.c]: Lines < 80 characters.

Comment (tadeu) [inline numpy/core/src/multiarray/arrayobject.c]: > Lines < 80 characters.

The largest line in the patch have 76 chars.
Or do you mean this line

`#define isspace(c)  ((c==' ')||(c=='\t')||(c=='\n')||(c=='\r')||(c=='\v')||(c=='\f'))`

which was already there?

**Comment by tadeu** [comment]
> This isn't actually a TST, that prefix is for official tests in tests directories.

Sorry, I've seem some TST commits with fixes for broken tests and assumed this would be too. What would be the correct prefix here?

**Comment by charris** [comment]
That would be a simple `BUG`.

File: numpy/core/src/multiarray/arrayobject.c (hunk 1 above)
Comment (charris) [inline numpy/core/src/multiarray/arrayobject.c]: Yep, might as well fix it.

**Comment by tadeu** [comment]
> That would be a simple `BUG`.

Ok, I've updated it.

**Comment by charris** [comment]
Hmm, there is a `NumPyOS_ascii_isspace` in `numpyos.{c,h}` that does the same thing. I wonder if we should be using that? There are several other uses of `isspace` that should perhaps be change.

**Comment by charris** [comment]
I'm wondering if we should make the macro definition unconditional? The other option is to use `NumPyOS_ascii_isspace` but change its signature to take an `int` instead of `char`.

**Comment by tadeu** [comment]
Ok, so what do you suggest? Something like this?

```
NPY_NO_EXPORT int
-NumPyOS_ascii_isspace(char c)
+NumPyOS_ascii_isspace(int c)
{
    return c == ' ' || c == '\f' || c == '\n' || c == '\r' || c == '\t'
                    || c == '\v';
}
```

```
-#define isspace(c)  ((c==' ')||(c=='\t')||(c=='\n')||(c=='\r')||(c=='\v')||(c=='\f'))
+#define isspace(c)  NumPyOS_ascii_isspace(c)
```

**Comment by charris** [comment]
I think the change to `int` for the argument takes care of ucs4, I don't think we run on any platforms where `int` isn't at least four bytes, so type promotion should be automatic. There are only a few uses of `isspace` in the file, just the two routines, so I'd get rid of the define and use the function directly. The ascii spaces are required for ucs4 in any case.

**Comment by charris** [comment]
Maybe want to use a cast to int for the (unsigned` ucs4. My machine has 64 bit ints, so no problem, but might be needed for machines with 32 bit ints.

**Comment by tadeu** [comment]
Should I change the other `NumPyOS_ascii_*` functions too? (e.g., `NumPyOS_ascii_isalpha`, `NumPyOS_ascii_isdigit`, etc.)

**Comment by tadeu** [comment]
I've updated the commit

**Comment by tadeu** [comment]
There's this error now:

``` 
ImportError: /home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/core/multiarray.so: undefined symbol: NumPyOS_ascii_isdigit
```

Isn't `numpyos.c` being linked into `multiarray`?

**Comment by charris** [comment]
> Isn't numpyos.c being linked into multiarray?

What happened to this comment?

**Comment by tadeu** [comment]
>>
>> Isn't numpyos.c being linked into multiarray?
>
> What happened to this comment?

Sorry, I deleted it, it was a mistake I made and quickly corrected (was
calling the wrong function) :)

**Comment by charris** [comment]
OK, let's give this a shot. I think it should be harmless (knock, knock). I'm not sure where all the other problems with the debug version come from. There is a special Python2.7 msvc compiler, msvc for python or some such,  that should be compatible with 2.7

**Comment by charris** [comment]
Thanks @tadeu .

**Comment by tadeu** [comment]
Thanks @charris !
Yes, unfortunatelly, I'm currently stuck using a version compiled with MSVC 2010 (and this means all other extensions too), because I'm interacting with other C++ code that depends on this compiler.
But if the other debug problems persist, I'll try to see if I can reproduce them with the vs2008 standard version.

**Pull request merged**